query-planner-js/src/buildPlan.ts

```diff
@@ -3436,42 +3451,11 @@ function addTypenameFieldForAbstractTypesInNamedFragments(fragments: NamedFragme
   //  1. expand any nested fragments in its selection.
   //  2. add `__typename` where we should in that expanded selection.
   //  3. re-optimize all fragments (using the "updated-with-typename" versions).
-  // which ends up getting us what we need. However, doing so requires us to deal with fragments in order
-  // of dependencies (first the ones with no nested fragments, then the one with only nested fragments of
-  // that first group, etc...), and that's why this method is a bit longer that one could have expected.
-  type FragmentInfo = {
-    original: NamedFragmentDefinition,
-    expandedSelectionSet: SelectionSet,
-    dependentsOn: string[],
-  };
-  const fragmentsMap = new Map<string, FragmentInfo>();
-
-  for (const fragment of fragments.definitions()) {
-    const expandedSelectionSet = addTypenameFieldForAbstractTypes(fragment.selectionSet.expandAllFragments());
-    const otherFragmentsUsages = new Map<string, number>();
-    fragment.collectUsedFragmentNames(otherFragmentsUsages);
-    fragmentsMap.set(fragment.name, {
-      original: fragment,
-      expandedSelectionSet,
-      dependentsOn: Array.from(otherFragmentsUsages.keys()),
-    });
-  }
-
-  const optimizedFragments = new NamedFragments();
-  while (fragmentsMap.size > 0) {
-    for (const [name, info] of fragmentsMap) {
-      // Note that graphQL specifies that named fragments cannot have cycles (https://spec.graphql.org/draft/#sec-Fragment-spreads-must-not-form-cycles)
-      // and so we guaranteed that on every ieration, at least element of the map is removed and thus that the
-      // overall `while` loops terminate.
-      if (info.dependentsOn.every((n) => optimizedFragments.has(n))) {
-        const reoptimizedSelectionSet = info.expandedSelectionSet.optimize(optimizedFragments);
-        optimizedFragments.add(info.original.withUpdatedSelectionSet(reoptimizedSelectionSet));
-        fragmentsMap.delete(name);
-      }
-    }
-  }
-
-  return optimizedFragments;
+  // which is what `mapToExpandedSelectionSets` gives us.
+  assert(!fragments.isEmpty(), 'Should not pass empty fragments to this method');
+  const updated = fragments.mapToExpandedSelectionSets((s) => addTypenameFieldForAbstractTypes(s));
```

Comment:
Extraneous intermediate function
```suggestion
  const updated = fragments.mapToExpandedSelectionSets(addTypenameFieldForAbstractTypes);
```